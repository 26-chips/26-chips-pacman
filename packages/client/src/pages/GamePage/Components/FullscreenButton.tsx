import { useEffect, useState } from 'react';
import styles from './gameBlock.module.scss';
import expand from 'assets/icons/fullscreen-expand.svg';
import shrink from 'assets/icons/fullscreen-shrink.svg';

export function FullscreenButton(): JSX.Element {
  const [isFullscreen, setFullscreen] = useState(false);
  const isFullscreenDisabled = !document.fullscreenEnabled;

  const toggleFullscreen = () => {
    if (isFullscreenDisabled) {
      alert('Fullscreen API not supported');
      return;
    }

    if (!document.fullscreenElement) {
      document.body.requestFullscreen();
    } else {
      document.exitFullscreen();
    }

    setFullscreen(Boolean(document.fullscreenElement));
  };

  useEffect(() => {
    const handleFSKey = (e: KeyboardEvent) => {
      if (e.key === '27') {
        e.preventDefault();
        toggleFullscreen();
      }
    };

    const setFS = () => {
      setFullscreen(Boolean(document.fullscreenElement));
    };

    document.addEventListener('keydown', handleFSKey);
    document.addEventListener('fullscreenchange', setFS);

    return () => {
      window.removeEventListener('keydown', handleFSKey);
      window.removeEventListener('fullscreenchange', setFS);
    };
  }, []);

  return (
    <img
      onClick={() => toggleFullscreen()}
      className={styles.fullscreenIcon}
      src={isFullscreen ? expand : shrink}
      alt="fullscreen switch"
    />
  );
}
