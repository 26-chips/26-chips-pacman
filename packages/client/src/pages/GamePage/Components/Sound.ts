import crackSound from 'assets/sounds/crack.mp3';
import gameLoopSound from 'assets/sounds/game_loop.mp3';
import failSound from 'assets/sounds/fail.mp3';
import countdownSound from 'assets/sounds/countdown.mp3';
import winSound from 'assets/sounds/win.mp3';
import bonusSound from 'assets/sounds/bonus.mp3';
import toggleSound from 'assets/sounds/toggle.mp3';
import blockedSound from 'assets/sounds/blocked.mp3';

class Sound {
  private soundUrl: string;

  private isLooped: boolean;

  private isMuted: boolean;

  private volumeLevel: number;

  private playbackRate: number;

  private playMultipleTimes: boolean;

  private audio: HTMLAudioElement | null;

  constructor(
    soundUrl: string,
    isLooped = false,
    volumeLevel = 1,
    playbackRate = 1,
    playMultipleTimes = false
  ) {
    if (soundUrl) {
      this.audio = new Audio(soundUrl);
      this.audio.loop = isLooped;
      this.audio.volume = volumeLevel;
      this.audio.playbackRate = playbackRate;
    } else {
      this.audio = null;
    }

    this.soundUrl = soundUrl;
    this.isMuted = false;
    this.isLooped = isLooped;
    this.volumeLevel = volumeLevel;
    this.playbackRate = playbackRate;
    this.playMultipleTimes = playMultipleTimes;
  }

  play = () => {
    if (!this.isMuted && this.audio) {
      this.audio.play();
      // play again if already playing
      if (!this.audio.paused && !this.audio.loop && this.playMultipleTimes) {
        // @ts-ignore
        this.audio.cloneNode(true).play();
      }
    }
  };

  pause = () => {
    if (this.audio) {
      this.audio.pause();
    }
  };

  toggleMute = () => {
    this.isMuted = !this.isMuted;

    if (this.isMuted) {
      // @ts-ignore
      this.audio.src = '';
    } else {
      // @ts-ignore
      this.audio.src = this.soundUrl;
    }
  };

  isPaused = () => {
    return this.audio?.paused;
  };
}

const win = new Sound(winSound, false, 0.5);
const fail = new Sound(failSound, false, 0.2);
const bonus = new Sound(bonusSound, false, 0.5);
const gameLoop = new Sound(gameLoopSound, true, 0.1);
const countdown = new Sound(countdownSound, false, 0.5);
const crack = new Sound(crackSound, false, 0.8, 0.8, true);
const blocked = new Sound(blockedSound, false, 0.4);
const toggle = new Sound(toggleSound, false, 0.2, 0.5);

export const AudioElements = {
  win: win,
  fail: fail,
  crack: crack,
  bonus: bonus,
  toggle: toggle,
  blocked: blocked,
  gameLoop: gameLoop,
  countdown: countdown,
};

export const toggleMute = () => {
  Object.values(AudioElements).forEach(el => {
    el.toggleMute();
  });
};

export const pauseAll = () => {
  Object.values(AudioElements).forEach(el => {
    el.pause();
  });
};
