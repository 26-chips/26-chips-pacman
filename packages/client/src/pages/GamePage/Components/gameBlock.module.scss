@import 'style_constants';

.canvas-container {
  margin-top: -200px;
  width:100%;
  text-align: center;

  &.paused {
    filter: blur(3px);
  }
}

.overlay {
  font-family: LexendPeta, sans-serif;
  font-weight: 300;
  position: fixed;
  inset: 0;
  display: flex;
  background-color: rgba(10 7 8 / 50%);
}

.overlay-content {
  margin: auto;
  color: $white;
}

.countdown {
  font-size: 260px;
}

.pause {
  font-size: 140px;
}

.game-control {
  display: flex;
  flex-direction: column;
  row-gap: 15px;
  border: 1px solid $blue;
  padding: 27px;
  border-radius: 8px;
  max-width: 250px;
  margin-top: 80px;
  margin-left: 60px;
}

.game-info {
  font-size: 20px;
  margin: 0;

  &-count {
    font-family: LexendPeta, sans-serif;
    font-weight: 700;
    margin-left: 10px;
  }
}
