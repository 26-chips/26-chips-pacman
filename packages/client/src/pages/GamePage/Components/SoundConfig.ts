import crackSound from 'assets/sounds/crack.mp3';
import gameLoopSound from 'assets/sounds/game_loop.mp3';
import failSound from 'assets/sounds/fail.mp3';
import winSound from 'assets/sounds/win.mp3';
import bonusSound from 'assets/sounds/bonus.mp3';
import toggleSound from 'assets/sounds/toggle.mp3';
import { Sound } from './Sound';

const win = new Sound(winSound, false, 0.5);
const fail = new Sound(failSound, false, 0.2);
const bonus = new Sound(bonusSound, false, 0.5);
const gameLoop = new Sound(gameLoopSound, true, 0.1);
const crack = new Sound(crackSound, false, 0.8, 0.8, true);
const toggle = new Sound(toggleSound, false, 0.2, 0.5);

export const AudioElements = {
  win: win,
  fail: fail,
  crack: crack,
  bonus: bonus,
  toggle: toggle,
  gameLoop: gameLoop,
};

export const toggleMute = () => {
  Object.values(AudioElements).forEach(el => {
    el.toggleMute();
  });
};

export const pauseAll = () => {
  Object.values(AudioElements).forEach(el => {
    el.pause();
  });
};
