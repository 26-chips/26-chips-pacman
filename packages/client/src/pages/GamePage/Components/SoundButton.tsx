import styles from './gameBlock.module.scss';
import muteIcon from 'assets/icons/mute.svg';
import playIcon from 'assets/icons/play.svg';
import { useState } from 'react';
import { AudioElements } from './SoundConfig';

interface SoundButtonProps {
  toggleMute: () => void;
}

export const SoundButton = (props: SoundButtonProps) => {
  const { toggleMute } = props;
  const [isMuted, setIsMuted] = useState(false);

  return (
    <img
      onClick={() => {
        toggleMute();
        setIsMuted(!isMuted);
        AudioElements.toggle.play();
      }}
      className={styles.icon}
      src={isMuted ? playIcon : muteIcon}
      alt="mute sound switch"
    />
  );
};
