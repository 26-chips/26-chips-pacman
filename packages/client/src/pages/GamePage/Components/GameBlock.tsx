import { ROUTES } from 'router';
import { CanvasComponent } from 'components/Canvas/Canvas';
import styles from './gameBlock.module.scss';
import { useState, useEffect, useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import { EndGameScreen } from 'components';
import cn from 'classnames';
import { FullscreenButton } from './FullscreenButton';
import { useFetchUserQuery, useAddUserToLeaderboardMutation } from 'api';
import { formUserName } from 'utils/helpers';
import { SoundButton } from './SoundButton';
import { AudioElements, pauseAll, toggleMute } from './SoundConfig';
import { GameResultForLeaderboard } from '../types';

const START_COUNT = 3;
const LIVES = 3;

export function GameBlock(): JSX.Element {
  const { data: user } = useFetchUserQuery();
  const navigate = useNavigate();

  const [points, setPoints] = useState(0);
  const [lives, setLives] = useState<number>(LIVES);
  const [time, setTime] = useState<number>(0);
  const [count, setCount] = useState(START_COUNT);
  const [isPaused, setIsPaused] = useState(true);
  const [maximumPoints, setMaximumPoints] = useState(0);

  const location = useLocation();

  // pause all sounds when exit /game
  useEffect(() => {
    pauseAll();
  }, [location]);

  useEffect(() => {
    if (isPaused) {
      AudioElements.gameLoop.pause();
    } else {
      AudioElements.gameLoop.play();
    }
  }, [isPaused]);

  const [addUserToLeaderboard] = useAddUserToLeaderboardMutation();

  const reduceLives = () => {
    setLives(prev => prev - 1);
  };

  const isCountDown = useMemo(() => {
    return count > 0;
  }, [count]);

  const allPillsCollected = useMemo(() => {
    return points > 0 && points === maximumPoints;
  }, [points, maximumPoints]);

  const gameIsOver = useMemo(() => {
    return lives <= 0 || allPillsCollected;
  }, [lives, maximumPoints, points]);

  const resetCounter = () => {
    setCount(START_COUNT);
  };

  // количество очков = собранные таблетки * на оставшиеся жизни + 5*(60 - потрченное время)
  const totalScore = useMemo(() => {
    const timeBonus = 5 * (60 - time) > 0 ? 5 * (60 - time) : 0;
    return points * (lives + 1) + timeBonus;
  }, [points, lives, time]);

  useEffect(() => {
    if (lives && !allPillsCollected) {
      const timer = count > 0 && setInterval(() => setCount(count - 1), 1000);
      return () => {
        if (timer) {
          clearInterval(timer);
        }
      };
    }
  }, [count, allPillsCollected]);

  useEffect(() => {
    if (count <= 0) {
      setIsPaused(false);
    }
  }, [count]);

  useEffect(() => {
    if (allPillsCollected) {
      setIsPaused(true);
    }
  }, [allPillsCollected]);

  // TODO найти способ получше задать черный фон
  useEffect(() => {
    document.body.classList.add('blackBG');
    return () => document.body.classList.remove('blackBG');
  });

  useEffect(() => {
    const submitData: GameResultForLeaderboard = {
      data: {
        points: totalScore,
        time: time,
        userId: user?.id,
        userNickname: formUserName(user, 'Guest'),
        userAvatar: user?.avatar,
      },
      ratingFieldName: 'points',
      teamName: 'chips',
    };

    const addResultToLeaderboard = async () => {
      try {
        await addUserToLeaderboard(submitData).unwrap();
      } catch (err) {
        throw new Error((err as Error).message);
      }
    };

    if (gameIsOver) {
      addResultToLeaderboard();
    }
  }, [gameIsOver]);

  const handleModalClose = () => {
    navigate(ROUTES.MAIN);
  };

  return (
    <>
      <div className={styles.gameControl}>
        <p className={styles.gameInfo}>
          Очки:
          <span className={styles.gameInfoCount}>{points}</span>
        </p>
        <p className={styles.gameInfo}>
          Оставшиеся жизни:
          <span className={styles.gameInfoCount}>{lives}</span>
        </p>
        <p className={styles.gameInfo}>
          Время:
          <span className={styles.gameInfoCount}>{time}</span>
        </p>
      </div>
      <div className={styles.iconContainer}>
        <FullscreenButton />
        <SoundButton toggleMute={toggleMute} />
      </div>
      <div className={cn(styles.canvasContainer, isPaused && styles.paused)}>
        {
          <CanvasComponent
            setPoints={setPoints}
            reduceLives={reduceLives}
            setTime={setTime}
            isPaused={isPaused}
            setIsPaused={setIsPaused}
            isCountDown={isCountDown}
            resetCounter={resetCounter}
            setMaximumPoints={setMaximumPoints}
            allPillsCollected={allPillsCollected}
          />
        }
      </div>

      <EndGameScreen
        className={styles.endGame}
        username={formUserName(user, 'Guest')}
        show={gameIsOver}
        onClose={handleModalClose}
        score={totalScore}
        elapsedTimeSec={time}
      />

      {isCountDown && (
        <div className={styles.overlay}>
          <div className={cn(styles.countdown, styles.overlayContent)}>
            {count}
          </div>
        </div>
      )}
      {isPaused && !isCountDown && (
        <div className={styles.overlay}>
          <div className={cn(styles.pause, styles.overlayContent)}>PAUSED</div>
        </div>
      )}
    </>
  );
}
