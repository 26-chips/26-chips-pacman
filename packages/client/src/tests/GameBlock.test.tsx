import { GameBlock } from 'pages/GamePage/Components/GameBlock';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { create } from 'react-test-renderer';

const mockComponent = (
  <BrowserRouter>
    <GameBlock />
  </BrowserRouter>
);

describe('<GameBlock />', () => {
  it('Should render', () => {
    render(mockComponent);
  });

  it('Should contain canvas element', () => {
    const { container } = render(mockComponent);
    expect(container.querySelector('canvas')).toBeInTheDocument();
  });

  it('Should match snapshot', () => {
    const tree = create(mockComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
