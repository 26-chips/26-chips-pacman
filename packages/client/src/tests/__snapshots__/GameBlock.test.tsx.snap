// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameBlock /> Should match snapshot 1`] = `
Array [
  <div>
    <p>
      Current points 0
    </p>
    <p>
      Current lives 3
    </p>
    <p>
      Time 0
    </p>
  </div>,
  <div
    className=""
  >
    <canvas />
  </div>,
  <div>
    <div
      className=""
    >
      3
    </div>
  </div>,
]
`;
